### PR TITLE
add metrics tracking the V8 heap usage

### PR DESCRIPTION
**Pull request by Geal**
This adds new gauge metrics tracking V8 memory usage:
- `apollo.router.v8.heap.used`
- `apollo.router.v8.heap.total`

This uses code introduced in the router-bridge in https://github.com/apollographql/federation-rs/pull/558

Due to the way we release the router-bridge, this updates the federation version to 2.8.3

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: .changesets/feat_geal_v8_heap_statistics.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+### add metrics tracking the V8 heap usage ([PR #5781](https://github.com/apollographql/router/pull/5781))
+
+We add new gauge metrics tracking V8 memory usage:
+- `apollo.router.v8.heap.used`: heap memory used by V8, in bytes
+- `apollo.router.v8.heap.total`: total heap allocated by V8, in bytes
```
Comment (shorgi) [inline .changesets/feat_geal_v8_heap_statistics.md:5]:
```suggestion
### Add V8 heap usage metrics ([PR #5781](https://github.com/apollographql/router/pull/5781))

The router supports new gauge metrics for tracking heap memory usage of the V8 Javascript engine:
- `apollo.router.v8.heap.used`: heap memory used by V8, in bytes
- `apollo.router.v8.heap.total`: total heap allocated by V8, in bytes
```

**Review by shorgi — APPROVED** [review]
docs LGTM

**Review by BrynCooke — APPROVED** [review]
Looks great.
@shorgi Could we look at separately creating some sort of troubleshooting docs that talk about when this is useful?

File: apollo-router/src/query_planner/bridge_query_planner_pool.rs [hunk 2]
```diff
@@ -119,21 +127,64 @@ impl BridgeQueryPlannerPool {
...
+            _v8_heap_total_gauge,
         })
     }
 
+    fn create_heap_used_gauge(meter: &Meter) -> (Arc<AtomicU64>, ObservableGauge<u64>) {
+        let current_heap_used = Arc::new(AtomicU64::new(0));
+        let current_heap_used_for_gauge = current_heap_used.clone();
+        let heap_used_gauge = meter
+            .u64_observable_gauge("apollo.router.v8.heap.used")
+            .with_description("V8 heap used, in bytes")
```
Comment (bnjjj) [inline apollo-router/src/query_planner/bridge_query_planner_pool.rs:167]: Could you also set a unit. Sometimes in some apms it's required to display the metrics properly

File: apollo-router/src/query_planner/bridge_query_planner_pool.rs [hunk 3]
```diff
@@ -119,21 +127,64 @@ impl BridgeQueryPlannerPool {
...
+            .init();
+        (current_heap_used, heap_used_gauge)
+    }
+
+    fn create_heap_total_gauge(meter: &Meter) -> (Arc<AtomicU64>, ObservableGauge<u64>) {
+        let current_heap_total = Arc::new(AtomicU64::new(0));
+        let current_heap_total_for_gauge = current_heap_total.clone();
+        let heap_total_gauge = meter
+            .u64_observable_gauge("apollo.router.v8.heap.total")
+            .with_description("V8 heap total, in bytes")
```
Comment (bnjjj) [inline apollo-router/src/query_planner/bridge_query_planner_pool.rs:180]: Same for the unit

File: apollo-router/src/query_planner/bridge_query_planner_pool.rs [hunk 4]
```diff
@@ -119,21 +127,64 @@ impl BridgeQueryPlannerPool {
             });
         }
         let sender_for_gauge = sender.clone();
-        let pool_size_gauge = meter_provider()
-            .meter("apollo/router")
+        let meter = meter_provider().meter("apollo/router");
+        let pool_size_gauge = meter
             .u64_observable_gauge("apollo.router.query_planning.queued")
```
Comment (bnjjj) [inline apollo-router/src/query_planner/bridge_query_planner_pool.rs:132]: I know it existed before but could you add description and unit please ?

**Review by bnjjj — CHANGES_REQUESTED** [review]

**Review by o0Ignition0o — APPROVED** [review]

File: apollo-router/src/query_planner/bridge_query_planner_pool.rs (hunk 4 above)
Comment (Geal) [reply apollo-router/src/query_planner/bridge_query_planner_pool.rs:132]: what kind of unit do we set here? Is there a convention for the expected units in otel?

Comment (Geal) [reply apollo-router/src/query_planner/bridge_query_planner_pool.rs:132]: (asking because I don't see many examples of setting units in our metrics)

**Comment by Geal** [comment]
@bnjjj is that the correct way to set units? [9e710a5](https://github.com/apollographql/router/pull/5781/commits/9e710a5e093e0f7114d66b7938ff23ddcf379883)

File: apollo-router/src/query_planner/bridge_query_planner_pool.rs (hunk 4 above)
Comment (bnjjj) [reply apollo-router/src/query_planner/bridge_query_planner_pool.rs:133]: Yes there are. At different places but for example for memory you can find an example here https://opentelemetry.io/docs/specs/semconv/system/hardware-metrics/#hwmemory---memory-module-metrics and more generally you can take examples from [here](https://opentelemetry.io/docs/specs/semconv/general/metrics/) there are some conventions too

**Comment by bnjjj** [comment]
@Geal yes except that for bytes it's usually `By` Check example [here](https://opentelemetry.io/docs/specs/semconv/system/hardware-metrics/#hwmemory---memory-module-metrics)

**Comment by Geal** [comment]
@bnjjj ok, it's fixed!

File: apollo-router/src/query_planner/bridge_query_planner_pool.rs (hunk 4 above)
Comment (bnjjj) [reply apollo-router/src/query_planner/bridge_query_planner_pool.rs:133]: For this on I would suggest using `query`

Comment (Geal) [reply apollo-router/src/query_planner/bridge_query_planner_pool.rs:133]: done

**Review by bnjjj — APPROVED** [review]

**Pull request merged**